File: puzzles/kotlin/src/chuck-norris.kt
import java.util.Scanner

fun main(args : Array<String>) {
  val input = Scanner(System.`in`)
  val message = input.nextLine()
  val binary = toBinary(message)
  var unary = toUnary(binary)
  println(unary)
}

fun toBinary(text: String): String {
  var binary = StringBuilder()
  for (character in text) {
    val binaryStr = Integer.toBinaryString(character.toInt())
    val number = binaryStr.toInt()
    binary.append(String.format("%07d", number))
  }
  return binary.toString()
}

fun toUnary(text: String): String {
  val unary = StringBuilder()
  var prevDigit = false // false = 0, true = 1
  if (text.isNotEmpty()) {
    val character = text[0]
    if (character == '0') {
      unary.append("00 0")
    } else {
      unary.append("0 0")
      prevDigit = true
    }
  }

  (1 until text.length)
      .asSequence()
      .map { text[it] }
      .forEach {
        if (it == '0' && prevDigit) {
          unary.append(" 00 0") // switch from 1 to 0
          prevDigit = false
        } else if (it == '1' && !prevDigit) {
          unary.append(" 0 0") // switch from 0 to 1
          prevDigit = true
        } else {
          unary.append("0") // repeat digit
        }
      }

  return unary.toString()
}

File: puzzles/python3/chuck_norris.py
def to_binary(text):
    binary_text = ""
    for character in text:
        binary_text += "{0:{fill}7b}".format(ord(character), fill="0")
    return binary_text


def to_unary(text):
    unary_text = ""
    prev_digit = False  # False = 0, True = 1
    # handle first character
    if len(text) >= 1:
        character = text[0]
        if character == "0":
            unary_text += "00 0"
        else:
            unary_text += "0 0"
            prev_digit = True

    for i in range(1, len(text)):
        character = text[i]
        if character == "0" and prev_digit:
            unary_text += " 00 0"  # switch from 1 to 0
            prev_digit = False
        elif character == "1" and not prev_digit:
            unary_text += " 0 0"  # switch from 0 to 1
            prev_digit = True
        else:
            unary_text += "0"  # repeat digit
    return unary_text


if __name__ == "__main__":
    text = input()
    binary_text = to_binary(text)
    unary_text = to_unary(binary_text)

    print(unary_text)
